defi-pulse/src/data-providers/coinapi.ts

```diff
@@ -0,0 +1,42 @@
+import { Requester } from '@chainlink/external-adapter'
+import Decimal from 'decimal.js'
+import { IndexAsset } from '../adapter'
+import { util } from '@chainlink/ea-bootstrap'
+
+const getPriceData = async (symbol: string) => {
+  const url = `https://rest.coinapi.io/v1/exchangerate/${symbol}/USD`
+  const config = {
+    url,
+    params: {
+      apikey: util.getRequiredEnv('API_KEY'),
+    },
+  }
+  const response = await Requester.request(config)
+  return response.data
+}
+
+const calculateIndex = (index: IndexAsset[]): number => {
+  let value = new Decimal(0)
+  index.forEach((i) => {
+    const price = i.priceData && i.priceData.rate
+    if (price <= 0) {
+      throw Error('invalid price')
+    }
+    value = value.plus(new Decimal(i.units).times(new Decimal(price)))
+  })
+  return value.toNumber()
+}
```

Comment:
Rename and extract this method to make it reusable:

```suggestion
const calculateIndexValue = (index: Index, toAssetPrice: (IndexAsset) => number): number =>
  index
    .map(asset => { ...asset, price: toAssetPrice(asset) }) 
    .reduce((acc, val) => acc.plus(val.units).times(val.price)), new Decimal(0))
    .toNumber()
}
```

`toAssetPrice` mapper should get the price, and throw if invalid. But why the price is not already added to the asset at this point? I think this should be done as soon as we fetch the price data.

Reply:
You are completely right. I'll get into it